### PR TITLE
Fix #51 #52 #53 #54 #53

### LINKED ISSUE #51: Start/Stop Assurance Groups
We should add to the AG Module to capability to start/stop an online analysis. 

We should support these 3 use cases:
- Start 
- Run X times 
- Stop 

Note: This is a customer request

### LINKED ISSUE #52: nae_prechange: Support for XML file upload
Currently we only support JSON file format. 

We should also support XML. 

Note: This is a customer request

### LINKED ISSUE #53: nae_prechange: Verify Documentation is wrong
https://github.com/CiscoDevNet/ansible-nae/blob/master/plugins/modules/nae_prechange.py#L40

Verify Is either `true` or `false` the doc speaks of `present` and `query` this needs to be fixed

### LINKED ISSUE #54: nae_prechange: Should not crash if AG does not exist
If the AG does not exist we should gracefully fail.

### PR DESCRIPTION
**Pull request by camrossi**
I did quite a few changes let me know if you wanna sync up over the phone. 

I DID Not write integration tests for the new features that is still to be done should I open an issue for that?

Fix #51 
Fix #52
Fix #53
Fix #54

### CODE REVIEW

File: plugins/module_utils/nae.py [hunk 1]
```diff
@@ -324,11 +325,57 @@ def newOfflineAG(self):
...
+                run_iter = "null"
+            else:
+                run_iter = str(self.params.get('run_iter'))
+            url = 'https://%(host)s:%(port)s/nae/api/v1/config-services/analysis' % self.params
+
+            form = '''
+                    {
+                    "assurance_group_list": [
+                        {
+                        "uuid": "''' + self.params['fabric_id'] + '''"
```
Comment (lhercot) [inline plugins/module_utils/nae.py]: self.params.get('fabric_id')

File: plugins/module_utils/nae.py [hunk 2]
```diff
@@ -392,13 +439,19 @@ def show_pre_change_analyses(self):
         self.result['Analyses'] = result
         return result
 
-    def is_json(self, myjson):
+    def is_json(self, filename):
```
Comment (lhercot) [inline plugins/module_utils/nae.py]: should we keep the is_json function to take text input and create a is_json_file to do the file opening and test with is_json function?

Same for XML

File: plugins/modules/nae_ag.py [hunk 3]
```diff
@@ -119,6 +152,8 @@ def main():
         apic_password=dict(type='str', default="", no_log=True),
         online=dict(type='bool', default=False),
         validate_certs=dict(type='bool', default=False),
+        run=dict(type='bool', default=None),
+        run_iter=dict(type='int', default=None),
```
Comment (lhercot) [inline plugins/modules/nae_ag.py]: should we spell it out run_iterations to make it clearer?

File: plugins/modules/nae_prechange.py [hunk 4]
```diff
@@ -39,9 +39,9 @@
     aliases: [ descr ]
   verify:
     description:
-    - When used with C(present) this flag specify if pre-change analysis is made from cisco.aci collection
-    - output_path config dump.
-    - When used with C(query) this flag will wait to execute the query until the prechange status is COMPLETED.
+    - This parameter is used to change the behaviour of the C(present) and C(query) state
```
Comment (lhercot) [inline plugins/modules/nae_prechange.py]: states

**Review by lhercot — CHANGES_REQUESTED** [review]
Also fix following errors caught by Sanity:
ERROR: plugins/modules/nae_ag.py:47:55: trailing-whitespace: Trailing whitespace
ERROR: plugins/modules/nae_ag.py:188:38: trailing-whitespace: Trailing whitespace
ERROR: plugins/modules/nae_ag.py:47:56: W291: trailing whitespace
ERROR: plugins/modules/nae_ag.py:188:39: W291: trailing whitespace
ERROR: plugins/module_utils/nae.py:51:1: E302: expected 2 blank lines, found 1
ERROR: plugins/module_utils/nae.py:203:16: ansible-format-automatic-specification: Format string contains automatic field numbering specification
ERROR: plugins/module_utils/nae.py:212:33: E127: continuation line over-indented for visual indent
ERROR: plugins/module_utils/nae.py:213:33: E127: continuation line over-indented for visual indent
ERROR: plugins/module_utils/nae.py:214:33: E127: continuation line over-indented for visual indent
ERROR: plugins/module_utils/nae.py:351:13: E265: block comment should start with '# '
ERROR: plugins/module_utils/nae.py:353:37: E127: continuation line over-indented for visual indent
ERROR: plugins/module_utils/nae.py:354:37: E127: continuation line over-indented for visual indent
ERROR: plugins/module_utils/nae.py:355:37: E127: continuation line over-indented for visual indent
ERROR: plugins/module_utils/nae.py:365:0: trailing-whitespace: Trailing whitespace
ERROR: plugins/module_utils/nae.py:365:1: W293: blank line contains whitespace
ERROR: plugins/module_utils/nae.py:367:33: E128: continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:368:33: E128: continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:376:0: trailing-whitespace: Trailing whitespace
ERROR: plugins/module_utils/nae.py:376:1: W293: blank line contains whitespace
ERROR: plugins/module_utils/nae.py:449:5: E301: expected 1 blank line, found 0
ERROR: plugins/module_utils/nae.py:455:5: E301: expected 1 blank line, found 0
ERROR: plugins/module_utils/nae.py:730:9: E265: block comment should start with '# '
ERROR: plugins/module_utils/nae.py:731:36: trailing-whitespace: Trailing whitespace
ERROR: plugins/module_utils/nae.py:731:37: W291: trailing whitespace
ERROR: plugins/module_utils/nae.py:734:0: trailing-whitespace: Trailing whitespace
ERROR: plugins/module_utils/nae.py:734:1: W293: blank line contains whitespace
ERROR: plugins/module_utils/nae.py:740:17: E122: continuation line missing indentation or outdented
ERROR: plugins/module_utils/nae.py:741:0: trailing-whitespace: Trailing whitespace
ERROR: plugins/module_utils/nae.py:741:1: W293: blank line contains whitespace
ERROR: plugins/module_utils/nae.py:763:0: trailing-whitespace: Trailing whitespace
ERROR: plugins/module_utils/nae.py:763:1: W293: blank line contains whitespace
ERROR: plugins/module_utils/nae.py:1070:0: trailing-whitespace: Trailing whitespace
ERROR: plugins/module_utils/nae.py:1070:1: W293: blank line contains whitespace
ERROR: plugins/module_utils/nae.py:1091:0: trailing-whitespace: Trailing whitespace
ERROR: plugins/module_utils/nae.py:1091:1: W293: blank line contains whitespace
ERROR: plugins/module_utils/nae.py:1093:29: E128: continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1094:33: E128: continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1095:29: E128: continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1096:21: E128: continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1097:29: E128: continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1098:29: E128 continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1101:13: E303 too many blank lines (2)
ERROR: plugins/module_utils/nae.py:1108:33: E128 continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1109:33: E128 continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1110:33: E128 continuation line under-indented for visual indent
ERROR: plugins/module_utils/nae.py:1128:5: E301 expected 1 blank line, found 0

File: plugins/module_utils/nae.py (hunk 1 above)
Comment (camrossi) [reply plugins/module_utils/nae.py]: Sure

File: plugins/module_utils/nae.py (hunk 2 above)
Comment (camrossi) [reply plugins/module_utils/nae.py]: I actually already changed it and was gonna be part of the next commit :)

File: plugins/modules/nae_ag.py (hunk 3 above)
Comment (camrossi) [reply plugins/modules/nae_ag.py]: Sure

File: plugins/module_utils/nae.py (hunk 2 above)
Comment (lhercot) [reply plugins/module_utils/nae.py]: I do not see the next commit you are talking about so should we still make that change?

File: plugins/module_utils/nae.py [hunk 5]
```diff
@@ -392,9 +438,17 @@ def show_pre_change_analyses(self):
...
+        f = open(filename)
         try:
-            json.loads(myjson)
+            json.loads(f.read())
+        except ValueError:
+            return False
+        return True
+
+    def is_xml(self, filename):
```
Comment (lhercot) [inline plugins/module_utils/nae.py]: same remark as for is_json. Would rename to is_json_file if we keep this way.

**Review by lhercot — CHANGES_REQUESTED** [review]

**Comment by shrsr** [comment]
Re-open

**Pull request closed without merging**